[clang-tidy] Improved readability redundant casting with enums

felix642 opened the pull request:
Fixed false negatives with readability-redundant-casting when the underlying types are the same and the option IgnoreTypeAliases is set to true.

Fixes #111137

felix642 committed:
[clang-tidy] Improved readability redundant casting with enums

Fixed false negatives with readability-redundant-casting when the underlying
types are the same and the option IgnoreTypeAliases is set to true.

Fixes #111137

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -222,6 +222,12 @@ Changes in existing checks
 
+- Improved :doc:`readability-redundant-casting
+  <clang-tidy/checks/readability/redundant-casting>` check by fixing
+  false negatives related to ``enum`` when setting ``IgnoreTypeAliases``
```
```suggestion
  false negatives related to ``enum`` when setting `IgnoreTypeAliases`
```

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -222,6 +222,12 @@ Changes in existing checks
+- Improved :doc:`readability-redundant-casting
+  <clang-tidy/checks/readability/redundant-casting>` check by fixing
+  false negatives related to ``enum`` when setting ``IgnoreTypeAliases``
+  to true.
```
```suggestion
  to `true`.
```

5chmidti commented:
> Fixed false negatives with readability-redundant-casting when the underlying types are the same and the option IgnoreTypeAliases is set to true.

To me,

> Fixed false negatives with readability-redundant-casting when casting an enum to its underlying type and the option IgnoreTypeAliases is set to true.

sounds better as the commit description.

---

The CI failure is probably because the implicit type for `E3` is not `unsigned int` on Windows.

felix642 committed:
fixup! [clang-tidy] Improved readability redundant casting with enums

Code Review & Removed non-portable test

felix642 committed:
amend! [clang-tidy] Improved readability redundant casting with enums

[clang-tidy] Improved readability redundant casting with enums

Fixed false negatives with readability-redundant-casting when casting
an enum to its underlying type and the option IgnoreTypeAliases is set to true.

Fixes #111137

felix642 commented:
> sounds better as the commit description.
I agree, thank you for the suggestion.

>The CI failure is probably because the implicit type for E3 is not unsigned int on Windows.
You are right. I could not find a way to make this test platform independent. I would either have to include <type_traits> or add preprocessor definitions to check for the OS (which might also break in the future). So I have decided to remove this test, since it didn't really add any value to the test suite.

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -222,6 +222,12 @@ Changes in existing checks
+  false negatives related to ``enum`` when setting `IgnoreTypeAliases`
+  to `true`.
+
+
```
Excessive newline.

felix642 committed:
fixup! [clang-tidy] Improved readability redundant casting with enums

Removed excessive newline

5chmidti reviewed (approved):
LGTM

+- some people may want to control this behavior and keep the explicit cast for enums, WDYT?

felix642 commented:
> +- some people may want to control this behavior and keep the explicit cast for enums, WDYT?

I can see why some people might want to explicitly cast the enum to its underlying type.
I could add the option `WarnEnumCastToUnderlyingType` which could default to `true`.

HerrCai0907 reviewed (commented):
Is it really a false negative? In another word, does `enum E1 : char {};` mean `E1` is the type alias of `char`?
IMO is not. for me a new option is more suitable for this case.

5chmidti commented:
> Is it really a false negative? In another word, does `enum E1 : char {};` mean `E1` is the type alias of `char`?
> IMO is not. for me a new option is more suitable for this case.

I think it is a false negative in the sense that the cast is redundant, but it is a type conversion, so maybe the default should not warn on explicit casts from enums, and make the loss of explicit Information opt-in.

felix642 commented:
@5chmidti and @HerrCai0907 I was revisiting this PR and I just remembered that my changes will only warn if `IgnoreTypeAliases` is set to `true`. 
The description for this option is :
> when set to true, it will resolve all type aliases and operate on the underlying types.

which, from my understanding, should cover enums.
I don't think we should add a second option to cover enums. What do you think ?

HerrCai0907 commented:
I am not the expect of cpp standard, in my opinion, enum is not a type alias. it is a new type.

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/readability/redundant-casting.cpp:
```diff
@@ -221,3 +221,16 @@ void testRedundantDependentNTTPCasting() {
+enum class E2 : char {};
+
+void testEnum(E1 e1, E2 e2){
+  char a = static_cast<char>(e1);
```
That isn't a redundant cast. Such casts should be done in explicit way. In C++ is even better to use:
std::to_underlying since C++23 or std::underlying_type now.

PiotrZSL reviewed (changes requested):
Please implement this change under separate option like: `AllowEnumsBeRecognizedAs Aliases` or something similar.
Implicit conversion between enums and underlying type is not always welcome by developers.

felix642 commented:
Closing this PR since I no longer think this is an option that we need.
